src/app/clusters/bindings/PendingNotificationMap.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <assert.h>
+
+#include <app/clusters/bindings/PendingNotificationMap.h>
+#include <app/util/binding-table.h>
+
+namespace chip {
+
+int16_t PendingNotificationMap::FindLRUBindingEntryIndex()
+{
+    uint8_t bindingWithSamePeer[EMBER_BINDING_TABLE_SIZE];
+
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    {
+        EmberBindingTableEntry entry;
+        emberGetBinding(i, &entry);
+        if (entry.type != EMBER_UNICAST_BINDING)
+        {
+            continue;
+        }
+        bool foundSamePeer = false;
+        for (uint8_t j = 0; j < i; j++)
+        {
+            EmberBindingTableEntry checkEntry;
+            emberGetBinding(j, &checkEntry);
+            if (checkEntry.type == EMBER_UNICAST_BINDING && checkEntry.fabricIndex == entry.fabricIndex &&
+                checkEntry.nodeId == entry.nodeId)
+            {
+                foundSamePeer          = true;
+                bindingWithSamePeer[i] = j;
+                break;
+            }
+        }
+        if (!foundSamePeer)
+        {
+            bindingWithSamePeer[i] = i;
+        }
+    }
+
+    uint16_t lastAppear[EMBER_BINDING_TABLE_SIZE];
+    for (uint16_t & value : lastAppear)
+    {
+        value = UINT16_MAX;
+    }
+    uint16_t appearIndex = 0;
+    for (PendingNotificationEntry pendingNotification : *this)
+    {
+        lastAppear[bindingWithSamePeer[pendingNotification.mBindingEntryId]] = appearIndex;
```

Comment:
It's very non-obvious what this is doing and why.  Why does it matter which bindings for a peer come before others in the binding table (which this code somehow cares about)?  What notion of "LRU" is this trying to implement, exactly?

This needs either much better variable naming or much better comments or most likely both.  But I also suspect that the behavior this is implementing does not make sense.

Reply:
The `pendingNotificationMap` works as a FIFO queue. When we fail to allocate a CASEClient we need to release an active outgoing CASE connection so we need to release the LRU entry from the pending notifications. That means the first peer in the queue.

Reply:
OK.  Documenting _that_ would be very helpful in terms of understanding what this code is trying to accomplish.